hcl2template/addrs/plugin.go

```diff
@@ -0,0 +1,238 @@
+package addrs
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"golang.org/x/net/idna"
+)
+
+// Plugin encapsulates a single plugin type.
+type Plugin struct {
+	Type      string
+	Namespace string
+	Hostname  string
+}
+
+func (p Plugin) RealRelativePath() string {
+	ns := DefaultPluginNamespace
+	if p.Namespace != "" {
+		ns = p.Namespace
+	}
+	return "/" + ns + "/packer-plugin-" + p.Type
+}
+
+func (p Plugin) Parts() []string {
+	return []string{p.Hostname, p.Namespace, p.Type}
+}
+
+func (p Plugin) String() string {
+	return strings.Join(p.Parts(), "/")
+}
+
+// ForDisplay returns a user-friendly FQN string, simplified for readability. If
+// the plugin is using the default hostname, the hostname is omitted.
+func (p *Plugin) ForDisplay() string {
+	parts := []string{}
+	if p.Hostname != DefaultPluginHost {
```

Comment:
what if p.Hostname is empty?

Reply:
Nevermind, this looks like it is never empty because of validation below.

Reply:
Yes that should be the case !